### ISSUE TITLE
Reward is getting lower and lower with each epoch, What can be the issue in training?

### ISSUE THREAD

**Issue by sakinafatima** [huggingface/trl#695]
Hello,

I am trying to optimize a T5 fine-tuned model for text generation task. At the moment, I am using BLEU score (between two texts) as a reward function. Before the optimization with PPO, model is able to produce an average BLEU score of 35% however with ppo, after each epoch, the reward is reducing so far. What is something I am doing wrong or should look into as I am new to RL? as the goal of PPO is to improve the reward or atleast make it more than the original bleu score of 35% that we got before model was optimized with PPO. 
this is my code:

```from transformers import  AutoModelForSeq2SeqLM
#loading the fine-tuned model
active_model=AutoModelForSeq2SeqLMWithValueHead.from_pretrained('small_gen_clean_prem/')
ref_model = AutoModelForSeq2SeqLMWithValueHead.from_pretrained('small_gen_clean_prem/')

batch_size = 200
config = PPOConfig(
    batch_size=batch_size,
    learning_rate=1.41e-5,
    mini_batch_size=16,
    gradient_accumulation_steps=1 #if I set to more than 1, I get empty tensors error
)

ppo_trainer = PPOTrainer(config, active_model, ref_model, tokenizer)


generation_kwargs = {
    "min_length": -1,
    "top_k": 0.0,
    "top_p": 1.0,
    "do_sample": True,
    "pad_token_id": tokenizer.eos_token_id
}


output_min_length = 4
output_max_length = 512
output_length_sampler = LengthSampler(output_min_length, output_max_length)`


score_all=[]
for i in range(20):
        input_tensors=[]
        output_tensors=[]
        score_=[]
        for data in valid_dataset:
                query_txt =data['input']
                query_tensor = tokenizer.encode(query_txt, return_tensors="pt").to(device)
                input_tensors.append(query_tensor.squeeze(0))
                desired_txt = data['ground_truth']
                print('desired text\n:',desired_txt)
                response_tensor = ppo_trainer.generate([item for item in query_tensor], return_prompt=False,length_sampler=output_length_sampler, **generation_kwargs)
                response_txt = tokenizer.decode(response_tensor[0], skip_special_tokens=True, max_new_tokens=512)
                output_tensors.append(response_tensor[0].squeeze(0))
              


                score = sentence_bleu([response_txt.split(),desired_txt.split()])
                score_.append(score)

        reward = [torch.FloatTensor([score]) for score in score_]


        score_all.append(np.mean(score_))
    
        
        train_stats = ppo_trainer.step(input_tensors,output_tensors,reward)
```
In the graph attached, y-axis is average mean score in each epoch.

<img width="377" alt="scores_ppo" src="https://github.com/huggingface/trl/assets/25576435/a07c26d9-46a8-432e-bf07-60eaaa0aeedc">

**Comment by BillChan226** [comment]
There could be many reasons. Maybe it's the issue of your batch size, as before steps of 5.0(M?), the model is trying to gain an accurate estimation of the overall bleu score over the whole dataset. After sweeping across every query in the set, the accuracy is starting to rise.

**Comment by malik-ali** [comment]
I would suggest starting with one of the example notebooks (like gpt sentiment) that is known to work and swapping in pieces like your reward, model, and data one at a time. You will get a sense of where you training is failing.

Also as a small thing to try, I identified a really weird behavior with `return_prompt=False`. Try changing that to true and just slicing the returned tensors to remove the prompt (as in the original gpt sentimentexample). It shouldn't make a difference but it did for me (see #697)

**Comment by BillChan226** [comment]
> I would suggest starting with one of the example notebooks (like gpt sentiment) that is known to work and swapping in pieces like your reward, model, and data one at a time. You will get a sense of where you training is failing.
> 
> 
> 
> Also as a small thing to try, I identified a really weird behavior with `return_prompt=False`. Try changing that to true and just slicing the returned tensors to remove the prompt (as in the original gpt sentimentexample). It shouldn't make a difference but it did for me (see #697)
> 
> 

I believe what you suggest is the most efficient way to try to make it work. Currently I'm trying to set up the StackLLaMa RLHF script ti reproduce the result.

**Comment by BillChan226** [comment]
I try to refine a stackLLama on RLHF and also encountered this issue. See [#723](https://github.com/huggingface/trl/issues/723). Would appreciate any help!
